src/compiler/crystal/syntax/parser.cr

```diff
@@ -5640,6 +5644,12 @@ module Crystal
 
       token
     end
+
+    def new_temp_arg_name
+      name = "__arg#{@temp_arg_count}"
+      @temp_arg_count += 1
+      name
+    end
```

Comment:
Can/Should this check if the name already exists in the current scope and discard it? Say I do

```cr
def foo(__arg0, @foo)
end
```

Reply:
The convention is that variables that start with `__` are reserved for the compiler, so they shouldn't clash with user variables. If a user choose to name their variables `__argN`, they are warned that that's problematic (at least in my mind :-P).

I don't want to make things more complex than they should, at this point.

Reply:
Mmh, not asking to do this here, but do you think it would actually be possible to make trying to use these a parser error then? I don't think it's too uncommon for somebody to try to use such variable names, especially when dealing with macros.

Reply:
I'm not sure it can be done. Sometimes methods are passed to macros, and these are `to_s`'d, and then reparsed again... and that would error when it shouldn't.

Maybe it's something to consider in the future.

Note that right now the `__arg` or `__temp` hack is used all over the place in the compiler (like, when you do `if a() || b()`, this isn't something new. And so far it didn't cause any problem, so I guess it's fine.